homeassistant/components/datadog/__init__.py

```diff
@@ -43,63 +48,100 @@
+    entry.runtime_data = statsd_client
 
     initialize(statsd_host=host, statsd_port=port)
 
     def logbook_entry_listener(event):
-        """Listen for logbook entries and send them as events."""
         name = event.data.get("name")
         message = event.data.get("message")
 
-        statsd.event(
+        entry.runtime_data.event(
             title="Home Assistant",
-            text=f"%%% \n **{name}** {message} \n %%%",
+            message=f"%%% \n **{name}** {message} \n %%%",
             tags=[
                 f"entity:{event.data.get('entity_id')}",
                 f"domain:{event.data.get('domain')}",
             ],
         )
 
-        _LOGGER.debug("Sent event %s", event.data.get("entity_id"))
-
     def state_changed_listener(event):
-        """Listen for new messages on the bus and sends them to Datadog."""
         state = event.data.get("new_state")
-
         if state is None or state.state == STATE_UNKNOWN:
             return
 
-        states = dict(state.attributes)
         metric = f"{prefix}.{state.domain}"
         tags = [f"entity:{state.entity_id}"]
 
-        for key, value in states.items():
-            if isinstance(value, (float, int)):
-                attribute = f"{metric}.{key.replace(' ', '_')}"
+        for key, value in state.attributes.items():
+            if isinstance(value, (float, int, bool)):
                 value = int(value) if isinstance(value, bool) else value
-                statsd.gauge(attribute, value, sample_rate=sample_rate, tags=tags)
-
-                _LOGGER.debug("Sent metric %s: %s (tags: %s)", attribute, value, tags)
+                attribute = f"{metric}.{key.replace(' ', '_')}"
+                entry.runtime_data.gauge(
+                    attribute, value, sample_rate=sample_rate, tags=tags
+                )
 
         try:
             value = state_helper.state_as_number(state)
+            entry.runtime_data.gauge(metric, value, sample_rate=sample_rate, tags=tags)
         except ValueError:
-            _LOGGER.debug("Error sending %s: %s (tags: %s)", metric, state.state, tags)
-            return
+            pass
 
-        statsd.gauge(metric, value, sample_rate=sample_rate, tags=tags)
+    entry.async_on_unload(
+        hass.bus.async_listen(EVENT_LOGBOOK_ENTRY, logbook_entry_listener)
+    )
+    entry.async_on_unload(
+        hass.bus.async_listen(EVENT_STATE_CHANGED, state_changed_listener)
+    )
 
-        _LOGGER.debug("Sent metric %s: %s (tags: %s)", metric, value, tags)
+    return True
 
-    hass.bus.listen(EVENT_LOGBOOK_ENTRY, logbook_entry_listener)
-    hass.bus.listen(EVENT_STATE_CHANGED, state_changed_listener)
 
+async def async_unload_entry(hass: HomeAssistant, entry: DatadogConfigEntry) -> bool:
+    """Unload a Datadog config entry."""
+    runtime = entry.runtime_data
+    runtime.flush()
+    runtime.close_socket()
     return True
+
+
+async def validate_datadog_connection(client: DogStatsd) -> bool:
```

Comment:
Move this function to `config_flow.py` as that seems to be where it's being used.